java/vector/src/main/codegen/templates/UnionVector.java

```diff
@@ -269,12 +373,14 @@ public void clear() {
   public Field getField() {
     List<org.apache.arrow.vector.types.pojo.Field> childFields = new ArrayList<>();
     List<FieldVector> children = internalMap.getChildren();
-    int[] typeIds = new int[children.size()];
     for (ValueVector v : children) {
-      typeIds[childFields.size()] = v.getMinorType().ordinal();
```

Comment:
@jacques-n This is the code I am referring to that assumes type id is the value of the enum MinorType rather than typeIds in the schema.

Reply:
Per my comment above, this isn't a violation of the spec as I read it, it simply doesn't support all variations of the spec.

Reply:
The idea is that an implementation might send codes over the wire like

[5, 10, 15, 5, 10, 15]

and then the type codes tell you that 5 is the 1st type in the union, 10 the 2nd, etc.

It was my recollection during the early Arrow days that there was a desire for certain codes to have a fixed interpretation. e.g. "4" might always be "JSON_STRING" in some Arrow implementation

Reply:
Yes, my perspective on arrow types in the java layer has been that one should collapse to leaf nodes (in general, I've always been against the more complex arbritrary approach the spec allows). In the situation we usually see, it doesn't make sense to ever have duplicate types within a union and we wanted to have a simplified path for this scenario.

For example:

v1_record = {
  date
  record_id
  measure1
}

v2_record = {
  date
  record_id
  measure2
}

I would suggest using this representation:

record = {
  date
  record_id
  measure1
  measure2
}

This typically works well because initial union values usually come from separate stream (e.g. files) and processing is cleaner if we don't have indirect for common fields (a frequent case).

 I'll take a look at the changes. I want to understand if you're introducing indirection or not.